aten/src/ATen/native/mps/operations/Indexing.mm

```diff
@@ -28,127 +29,127 @@
 namespace at {
 namespace native {
 
-static
-void getDataOffsets(const TensorIteratorBase& iter, simd_uint3* data_offsets, uint32_t num_outputs) {
-  // Collect offsets for Input, Output and Index tensors
-  const uint32_t nOffsets = 3;
-  const uint32_t nDim = iter.ndim();
-
-  std::vector<uint32_t> strides[nDim];
-  const IntArrayRef& iterShape = iter.shape();
-
-  for (uint32_t i=0; i < nDim; i++)
-    for (uint32_t offset = 0; offset < nOffsets; offset++)
-      strides[i].push_back(iter.strides(offset)[i]);
-
-  for (int32_t i = 0; i < num_outputs; i++) {
-    uint32_t idx = i;
-    for (int dim = 0; dim < nDim; dim++) {
-      uint32_t remainder = idx % iterShape[dim];
-      idx /= iterShape[dim];
-
-      for (int offset = 0; offset < nOffsets; offset++) {
-        data_offsets[i][offset] += remainder * strides[dim][offset];
-      }
-    }
-  }
-}
-
 static
 bool dispatchIndexKernel(TensorIteratorBase& iter, IntArrayRef index_size, IntArrayRef index_stride, bool index_select, bool accumulate) {
   using namespace mps;
 
-  @autoreleasepool {
-    if (iter.numel() == 0) {
-      return true;
-    }
-
-    const Tensor& inputTensor = iter.tensor(1);
-    Tensor outputTensor = iter.tensor(0);
+  if (iter.numel() == 0)
+    return true;
 
-    id<MTLBuffer> inputBuffer  = getMTLBufferStorage(inputTensor);
-    id<MTLBuffer> outputBuffer = getMTLBufferStorage(outputTensor);
-    MTLResourceOptions options = [inputBuffer resourceOptions];
-    int64_t iter_numel = iter.numel();
-    int64_t num_indices = index_size.size();
-    NSError* error = nil;
+  const Tensor& inputTensor = iter.tensor(1);
+  Tensor outputTensor = iter.tensor(0);
+  id<MTLBuffer> inputBuffer  = getMTLBufferStorage(inputTensor);
+  id<MTLBuffer> outputBuffer = getMTLBufferStorage(outputTensor);
+  MTLResourceOptions options = [inputBuffer resourceOptions];
+  MPSStream* mpsStream = getCurrentMPSStream();
+  id<MTLDevice> device = MPSDevice::getInstance()->device();
+
+  dispatch_sync(mpsStream->queue(), ^(){
+    @autoreleasepool {
+      NSError* error = nil;
+      int64_t num_indices = index_size.size();
+      const uint32_t numThreads = iter.numel();
+      const uint32_t nOffsets = 3;
+      const uint32_t nDim = iter.ndim();
+      uint32_t strides[nDim][nOffsets];
+      const IntArrayRef& iterShape = iter.shape();
+      std::vector<uint32_t> iterShapeVec;
```

Comment:
Here and below, lets use the arrays `at::detail::Array` to avoid any C++ overhead.

Reply:
Sounds good - will update it